### PR TITLE
[SPARK-12799] Simplify various string output for expressions

### PR DESCRIPTION
**Pull request by liancheng**
This PR introduces several major changes:
1. Replacing `Expression.prettyString` with `Expression.sql`
   
   The `prettyString` method is mostly an internal, developer faced facility for debugging purposes, and shouldn't be exposed to users.
2. Using SQL-like representation as column names for selected fields that are not named expression (back-ticks and double quotes should be removed)
   
   Before, we were using `prettyString` as column names when possible, and sometimes the result column names can be weird.  Here are several examples:
   
   | Expression | `prettyString` | `sql` | Note |
   | --- | --- | --- | --- |
   | `a && b` | `a && b` | `a AND b` |  |
   | `a.getField("f")` | `a[f]` | `a.f` | `a` is a struct |
3. Adding trait `NonSQLExpression` extending from `Expression` for expressions that don't have a SQL representation (e.g. Scala UDF/UDAF and Java/Scala object expressions used for encoders)
   
   `NonSQLExpression.sql` may return an arbitrary user facing string representation of the expression.

### CODE REVIEW

**Comment by liancheng** [comment]
(I'd expect multiple test failures due to changes in exception error messages.)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 1]
```diff
@@ -233,7 +232,7 @@ private[hive] case class HiveGenericUDF(
   }
 
   @transient
-  private lazy val deferedObjects = argumentInspectors.zip(children).map { case (inspect, child) =>
+  private lazy val deferredObjects = argumentInspectors.zip(children).map { case (inspect, child) =>
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: The original field name is misspelled here.

**Comment by liancheng** [comment]
cc @rxin

**Comment by rxin** [comment]
High level changes LGTM

**Comment by rxin** [comment]
almost there...

one question: is ` a valid character for column name?

**Comment by rxin** [comment]
(I'm thinking we should remove all backticks when generating name aliases)

**Comment by liancheng** [comment]
I should add a `quote` method somewhere to handle backtick escaping (Hive escapes backticks within identifiers to double-backticks). By "name aliases", do you mean those like generated subquery names?

**Comment by rxin** [comment]
I was referring to the column names automatically generated by spark sql. It is just taken from the "sql" representation of the column names right now, which includes backticks and can be a problem.

**Comment by yhuai** [comment]
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala#L171 has the extra space.

**Comment by liancheng** [comment]
A summary of offline discussion with @rxin and @marmbrus:

The reason why @rxin suggested removing all back-ticks in generated column names is mostly because of backwards compatibility. Those generated column names are now generated using `Expression.sql` instead of `Expression.prettyString`. For example, the following DataFrame

```scala
df.selectExpr("id + 1")
```

used to produce a single column named `id + 1`, but now it becomes `` `id` + 1``.

However, later on I found that removing back-ticks in generated column names still cannot guarantee this level of backwards-compatibility. This is because, although `pettyString` and `sql` are often quite similar, they are still inherently different from each other in many cases, for example:

Expression | `prettyString` | `sql`
---------- | -------------- | -----
`a && b` | `a && b` | `a AND b`
`a.getField("f")` | `a[f]` | `a.f`
`m.getItem("key")` | `m[key]` | `m["key"]`

Basically we won't be able to replace `prettyString` with `sql` is we do want this level of backwards-compatibility.  However, we may see that the original `prettyString` method doesn't return proper column names.  I tend to fix them in Spark 2.0.  As for back-tick quoting, a utility method `safeSQLIdent` is added to quote identifiers only when necessary.

**Comment by rxin** [comment]
@liancheng why do we need the quote parameter for the sql function? Can't we just replace the attribute references with PrettyAttribute, which does not include backticks?

**Comment by marmbrus** [comment]
+1 to doing this with a transformation instead of threading a boolean all the way through.

**Comment by liancheng** [comment]
Ah, that's a good point! Totally forgot `PrettyAttribute`.

**Comment by rxin** [comment]
I'd actually not have "PythonUDF#" as part of the name.

**Comment by liancheng** [comment]
It took me sometime to fix those R test failures. Hopefully the last commit fixes everything (finger crossed).

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala [hunk 2]
```diff
@@ -131,17 +131,17 @@ abstract class OrcSuite extends QueryTest with TestHiveSingleton with BeforeAndA
...
+        |  CAST(null as SMALLINT) as c1,
+        |  CAST(null as INT) as c2,
+        |  CAST(null as BIGINT) as c3,
+        |  CAST(null as FLOAT) as c4,
+        |  CAST(null as DOUBLE) as c5,
+        |  CAST(null as DECIMAL(7,2)) as c6,
+        |  CAST(null as TIMESTAMP) as c7,
+        |  CAST(null as DATE) as c8,
+        |  CAST(null as STRING) as c9,
+        |  CAST(null as VARCHAR(10)) as c10
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: Aliases are added mostly for the last two fields: `VARCHAR` is equivalent to `STRING` for Spark SQL, so the last two fields actually have exactly the same column names.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 3]
```diff
@@ -39,11 +39,11 @@ case class ScalaUDF(
     dataType: DataType,
     children: Seq[Expression],
     inputTypes: Seq[DataType] = Nil)
-  extends Expression with ImplicitCastInputTypes {
+  extends Expression with ImplicitCastInputTypes with NonSQLExpression {
 
   override def nullable: Boolean = true
 
-  override def toString: String = s"UDF(${children.mkString(",")})"
+  override def toString: String = s"ScalaUDF(${children.mkString(",")})"
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Is this change necessary? It is really strange to call this ScalaUDF because Java also goes through this.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: We should rename the case class too - but that should be done separately.

Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Oh right, I didn't realize that Java UDF also goes here. Will revert this one.

**Comment by rxin** [comment]
A few comments:

1. As discussed offline, let's be consistent and always quote in sql query generation, and never allow backticks in aliases.

2. Rename safeSQLIdent -> quoteIdentifier

3. Please add a unit test case for each expression that requires safeSQLIdent

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 4]
```diff
@@ -298,7 +299,14 @@ case class AttributeReference(
 case class PrettyAttribute(name: String, dataType: DataType = NullType)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: We should update the classdoc for PrettyAttribute to reflect what it is doing.

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
@rxin Thanks for the review and the summary. Will finish them in a follow-up PR.

**Comment by liancheng** [comment]
@rxin Shall we merge this?

**Comment by rxin** [comment]
I don't think this is blocking other work or in a code that changes frequently, so I'd prefer we just fix all the issues and have proper test coverage before merging.

**Comment by rxin** [comment]
BTW You might want to resolve this before it goes super stale.

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 5]
```diff
@@ -141,11 +144,11 @@ case class UnresolvedFunction(
   override def nullable: Boolean = throw new UnresolvedException(this, "nullable")
   override lazy val resolved = false
 
-  override def prettyString: String = {
-    s"${name}(${children.map(_.prettyString).mkString(",")})"
+  override def sql: String = {
+    s"$name(${children.map(_.sql).mkString(", ")})"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Is this same of the default one in `Expression`? after we override `prettyName` with `name`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 6]
```diff
@@ -499,6 +489,8 @@ abstract class BinaryOperator extends BinaryExpression with ExpectsInputTypes {
 
   def symbol: String
 
+  def sqlOperator: String = symbol
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: why add this concept?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala [hunk 7]
```diff
@@ -123,4 +123,6 @@ case class BitwiseNot(child: Expression) extends UnaryExpression with ExpectsInp
   }
 
   protected override def nullSafeEval(input: Any): Any = not(input)
+
+  override def sql: _root_.scala.Predef.String = s"~${child.sql}"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala]: sometime IDE is stupid...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 8]
```diff
@@ -93,6 +93,8 @@ object ExtractValue {
   }
 }
 
+trait ExtractValue { this: Expression => }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: can just be `trait ExtractValue extends Expression`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 9]
```diff
@@ -144,10 +152,11 @@ case class GetArrayStructFields(
     field: StructField,
     ordinal: Int,
     numFields: Int,
-    containsNull: Boolean) extends UnaryExpression {
+    containsNull: Boolean) extends UnaryExpression with ExtractValue {
 
   override def dataType: DataType = ArrayType(field.dataType, containsNull)
-  override def toString: String = s"$child.${field.name}"
+  override def toString: String = s"$child[$ordinal].${field.name}"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: The javadoc of `GetArrayStructFields` is too simple...

Actually `GetArrayStructFields` means, we have an array, each element of this array is struct type, and for each element, we wanna get one field of it. It's like a batched version of `GetStructField`, just like `persons.map(_.name)` is a batched version of `person.name`.

So `s"$child.${field.name}"` is corrected here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 10]
```diff
@@ -310,10 +311,20 @@ case class AttributeReference(
  * A place holder used when printing expressions without debugging information such as the
  * expression id or the unresolved indicator.
  */
-case class PrettyAttribute(name: String, dataType: DataType = NullType)
+case class PrettyAttribute(
+                            name: String,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: fix indent

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 11]
```diff
@@ -130,19 +133,28 @@ package object util {
     ret
   }
 
+  // Replaces attributes, string literals, complex type extractors with their pretty form so that
+  // generated column names don't contain back-ticks or double-quotes.
+  def usePrettyExpression(e: Expression): Expression = e transform {
+    case a: Attribute => new PrettyAttribute(a)
+    case Literal(s: UTF8String, StringType) => PrettyAttribute(s.toString, StringType)
+    case e @ GetStructField(child, _, Some(name)) => PrettyGetStructField(child, name, e.dataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala]: can we also turn `GetStructField` to `PrettyAttribute` here? We can generate the sql string here and pass it as the name of `PrettyAttribute`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala [hunk 12]
```diff
@@ -36,9 +36,12 @@ case class PythonUDF(
...
-    children: Seq[Expression]) extends Expression with Unevaluable with Logging {
+    children: Seq[Expression])
+  extends Expression with Unevaluable with NonSQLExpression with Logging {
 
-  override def toString: String = s"PythonUDF#$name(${children.mkString(",")})"
+  override def toString: String = s"PythonUDF#$name(${children.mkString(", ")})"
 
   override def nullable: Boolean = true
+
+  override def sql: String = s"$name(${children.mkString(", ")})"
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala]: It extends `NonSQLExpression`, no need to implement it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 13]
```diff
@@ -255,6 +236,19 @@ trait Unevaluable extends Expression {
 
 
 /**
+ * Expressions that don't have SQL representation should extend this trait.  Examples are
+ * `ScalaUDF`, `ScalaUDAF`, and object expressions like `MapObjects` and `Invoke`.
+ */
+trait NonSQLExpression extends Expression {
+  override def sql: String = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: how about mark it as final?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 14]
```diff
@@ -37,11 +38,21 @@ import org.apache.spark.sql.execution.datasources.LogicalRelation
  * supported by this builder (yet).
  */
 class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Logging {
+  require(logicalPlan.resolved, "SQLBuilder only supports reslved logical query plans")
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: nit: `reslved` -> `resolved`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 6 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: For example, `And.symbol` is `&&`, while `And.sqlOperator` is `AND`. The former is used in Hive shims for converting Hive filters. The latter is used to generate SQL string.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala (hunk 5 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Yeah, I can override `prettyName` instead of `sql` here. Will update. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 6 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: makes sense, but do we override it somewhere?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala (hunk 7 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala]: Agree...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 8 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 9 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: Ah, nice catch! Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala (hunk 11 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala]: Probably not. Because identifiers in `GetStructField.sql` are quoted.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala (hunk 12 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala]: `Expression.sql` is also called when generating column names. `NonSQLExpressoin.sql` calls `PythonUDF.toString` here, which returns `PythonUDF#...`. We probably don't want to have `#` appear in column names.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 13 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: One exception is the `PythonUDF` case mentioned above.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala (hunk 14 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 6 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: It's overriden in `And` and `Or`.

Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Oh, although I added the method, I forgot to use it in `And` and `Or` in the revision you commented at first. Just fixed that in my last commit.

**Comment by liancheng** [comment]
@rxin I think this one is ready for sign off now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala (hunk 12 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala]: So it doesn't need to extend `NonSQLExpression`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 15]
```diff
@@ -130,12 +134,17 @@ case class GetStructField(child: Expression, ordinal: Int, name: Option[String]
       }
     })
   }
+}
+
+case class PrettyGetStructField(child: Expression, name: String, dataType: DataType)
+  extends UnaryExpression with Unevaluable {
 
-  override def sql: String = child.sql + s".`${childSchema(ordinal).name}`"
+  override def sql: String = s"${child.sql}.$name"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: can we use this string as the name of `PrettyAttribute`? Then we can avoid creating these classes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 16]
```diff
@@ -310,10 +311,20 @@ case class AttributeReference(
  * A place holder used when printing expressions without debugging information such as the
  * expression id or the unresolved indicator.
  */
-case class PrettyAttribute(name: String, dataType: DataType = NullType)
+case class PrettyAttribute(
+    name: String,
+    dataType: DataType = NullType,
+    override val foldable: Boolean = false)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: Is there some case we need the foldable to be true?

**Comment by cloud-fan** [comment]
My last concern is changing default column alias `_c0, _c1, ..` to `Expression.sql`. Before this PR, query plans generated by SQL string will use `_c0, _c1, ..`, query plans generated by DataFrame will use `prettyString`. Now we always use `Expression.sql`, this is more consistent, but will break some old code, e.g. `select _c1 from (select a + 1, b + 2 from tbl) t `, does it worth?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala (hunk 16 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: Oh, this is an unexpected change. I should revert it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 15 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: No, `child.sql` may contain back-ticks/double-quotes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 17]
```diff
@@ -198,18 +208,26 @@ case class GetArrayStructFields(
   }
 }
 
+case class PrettyGetArrayStructFields(
+    child: Expression, ordinal: Int, name: String, dataType: DataType)
+  extends UnaryExpression with Unevaluable {
+
+  override def sql: String = s"${child.sql}[$ordinal].$name"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: should be `s"${child.sql}.$name"`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 15 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: what if we use `transformUp` in `usePrettyExpression`, so that we can make sure when we generate this string, `child.sql` won't have back-ticks/double-quotes?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 17 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: Yes, thanks!

**Comment by liancheng** [comment]
@cloud-fan I'm open to the new column naming issue. Personally I tend to make it more consistent. Please note that replace `prettyString` with `sql` already breaks compatibility (e.g. `a && b` is renamed to `a AND b`).

**Comment by liancheng** [comment]
retest this please

**Comment by rxin** [comment]
I likely won't have time to look at this until next week. Feel free to merge if @cloud-fan lgtms. I can do some post-hoc review later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala (hunk 12 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDF.scala]: Well, it does. Because we can't generate SQL string for a query plan containing a Python UDF. `SQLBuilder` tries to find `NonSQLExpression`s for this detection.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 15 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: Good idea. Updated and removed these new classes.

**Comment by cloud-fan** [comment]
LGTM as long as the tests pass.

**Comment by liancheng** [comment]
Thanks for the detailed review!

**Comment by liancheng** [comment]
One more update after offline discussion with @rxin: numeric literals appearing in column names shouldn't come with numeric suffixes or casts. This is done by replacing numeric literals with `PrettyAttribute` in `usePrettyExpression`.

**Comment by liancheng** [comment]
OK I'm going to merge this to master.
